File: 21. Zigzag traversal.cpp
#include <bits/stdc++.h>
using namespace std;
struct Node
{
    int data;
    Node* left;
    Node* right;
    Node(int d) : data(d), left(nullptr), right(nullptr) {}
};
void zigzag(Node* root)
{
    if (!root)
        return;
    stack<Node*> curr, next;
    bool leftToRight = true;
    curr.push(root);
    while (!curr.empty())
    {
        Node* cur = curr.top();
        curr.pop();
        cout << cur->data << " ";
        if (leftToRight)
        {
            if (cur->left)
                next.push(cur->left);
            if (cur->right)
                next.push(cur->right);
        }
        else
        {
            if (cur->right)
                next.push(cur->right);
            if (cur->left)
                next.push(cur->left);
        }
        if (curr.empty())
        {
            leftToRight = !leftToRight;
            swap(curr, next);
        }
    }
    cout << endl;
}
int main()
{
    Node* root = new Node(1);
    root->left = new Node(2);
    root->right = new Node(3);
    root->left->left = new Node(4);
    root->left->right = new Node(5);
    cout << "Zigzag: ";
    zigzag(root);
    return 0;
}

File: 21. Zigzag traversal.py
from collections import deque
class Node:
    def __init__(self,data): self.data=data; self.left=None; self.right=None

def zigzag(root):
    if not root: return
    curr=[root]
    left_to_right=True
    while curr:
        next_level=[]
        for node in curr[::-1]:
            print(node.data, end=' ')
            if left_to_right:
                if node.left: next_level.append(node.left)
                if node.right: next_level.append(node.right)
            else:
                if node.right: next_level.append(node.right)
                if node.left: next_level.append(node.left)
        left_to_right=not left_to_right
        curr=next_level
    print()

if __name__=='__main__':
    root=Node(1); root.left=Node(2); root.right=Node(3); root.left.left=Node(4); root.left.right=Node(5)
    print('Zigzag:', end=' '); zigzag(root)
